src/Data/List/Properties.agda

```diff
@@ -122,28 +122,50 @@ map-injective finj {x ∷ xs} {y ∷ ys} eq =
 ------------------------------------------------------------------------
 -- mapMaybe
 
+length-catMaybes : ∀ xs → length (catMaybes′ {A = A} xs) ≤ length xs
+length-catMaybes []             = ≤-refl
+length-catMaybes (just x  ∷ xs) = s≤s (length-catMaybes xs)
+length-catMaybes (nothing ∷ xs) = m≤n⇒m≤1+n (length-catMaybes xs)
+
+catMaybesTest : catMaybes {A = A} ≗ catMaybes′
+catMaybesTest []             = refl
+catMaybesTest (just x ∷ xs)  = cong (x ∷_) (catMaybesTest xs)
+catMaybesTest (nothing ∷ xs) = catMaybesTest xs
+
+mapMaybeTest : (p : A → Maybe B) → mapMaybe p ≗ mapMaybe″ p
+mapMaybeTest p []       = refl
+mapMaybeTest p (x ∷ xs) with ih ← mapMaybeTest p xs | p x
+... | nothing = ih
+... | just y  = cong (y ∷_) ih
+
 mapMaybe-just : (xs : List A) → mapMaybe just xs ≡ xs
 mapMaybe-just []       = refl
 mapMaybe-just (x ∷ xs) = cong (x ∷_) (mapMaybe-just xs)
 
+mapMaybe′-just : (xs : List A) → mapMaybe′ just xs ≡ xs
+mapMaybe′-just []       = refl
+mapMaybe′-just (x ∷ xs) = cong (x ∷_) (mapMaybe′-just xs)
+
 mapMaybe-nothing : (xs : List A) →
-                   mapMaybe {B = A} (λ _ → nothing) xs ≡ []
+                   mapMaybe (λ _ → nothing {A = A}) xs ≡ []
```

Comment:
Should this be generalised?

```suggestion
                   mapMaybe (λ _ → nothing {A = B}) xs ≡ []
```

Reply:
And latest commit restores the binding of `B` to `mapMaybe`...